src/main/scala/firrtl/transforms/InlineBooleanExpressions.scala

```diff
@@ -0,0 +1,152 @@
+    (info1, info2) match {
+      case (FileInfo(fileLineRegex(file1, line1)), FileInfo(fileLineRegex(file2, line2))) =>
+        (file1 == file2) && (line1 == line2)
+      case (MultiInfo(infos1), MultiInfo(infos2)) if infos1.size == infos2.size =>
+        infos1.zip(infos2).forall { case (i1, i2) =>
+          sameFileAndLineInfo(i1, i2)
+        }
+      case (NoInfo, NoInfo) => true
+      case _ => false
+    }
+  }
+
+  private class MapMethods(maxInlineCount: Int, dontTouches: Set[Ref]) {
+    val netlist: Netlist = new Netlist
+    val inlineCounts = mutable.Map.empty[Ref, Int]
+    var inlineCount: Int = 1
+
+    def onExpr(info: Info, outerExpr: Option[Expression])(expr: Expression): Expression = {
+      expr match {
+        case ref: WRef if !dontTouches.contains(ref.name.Ref) && ref.name.head == '_' =>
+          val refKey = ref.name.Ref
+          netlist.get(we(ref)) match {
+            case Some((refExpr, refInfo)) if sameFileAndLineInfo(info, refInfo) =>
+              val inlineNum = inlineCounts.getOrElse(refKey, 1)
+              if (!outerExpr.isDefined || canInline(outerExpr.get, ref, refExpr) && (inlineNum + inlineCount) <= maxInlineCount) {
+                inlineCount += inlineNum
+                refExpr
+              } else {
+                ref
+              }
+            case other => ref
+          }
+        case other => other.mapExpr(onExpr(info, Some(other)))
+      }
+    }
+
+    def onStmt(stmt: Statement): Statement = {
+      stmt.mapStmt(onStmt) match {
+        case hasInfo: HasInfo =>
+          inlineCount = 1
+          val stmtx = hasInfo.mapExpr(onExpr(hasInfo.info, None))
+          stmtx match {
+            case node: DefNode => inlineCounts(node.name.Ref) = inlineCount
+            case _ =>
+          }
+          stmtx match {
+            case node @ DefNode(info, name, value) =>
+              netlist(we(WRef(name))) = (value, info)
+            case _ =>
+          }
+          stmtx
+        case other => other
+      }
+    }
+  }
+
+  def execute(state: CircuitState): CircuitState = {
+    val dontTouchMap: Map[OfModule, Set[Ref]] = {
+      val refTargets = state.annotations.flatMap {
+        case anno: HasDontTouches => anno.dontTouches
+        case o => Nil
+      }
+      val dontTouches: Seq[(OfModule, Ref)] = refTargets.map {
+        case r => Target.referringModule(r).module.OfModule -> r.ref.Ref
+      }
+      dontTouches.groupBy(_._1).mapValues(_.map(_._2).toSet).toMap
+    }
+
+    val maxInlineCount = state.annotations.collectFirst {
+      case InlineBooleanExpressionsMax(max) => max
+    }.getOrElse(InlineBooleanExpressions.defaultMax)
+
+    val modulesx = state.circuit.modules.map { m =>
+      val mapMethods = new MapMethods(maxInlineCount, dontTouchMap.getOrElse(m.name.OfModule, Set.empty[Ref]))
+      m.mapStmt(mapMethods.onStmt(_))
+    }
+
+    state.copy(circuit = state.circuit.copy(modules = modulesx))
```

Comment:
Remove InlineBoioleanExpressionsMax annotations from the returned state.

Reply:
What if for some reason the transform gets invalidated and needs to rerun? the second run would no longer have access to `InlineBooleanExpressionsMax`. I'm fine with removing it, just wondering what the behavior should be in this case.